src/disco/bank/fd_txncache.c

```diff
@@ -0,0 +1,741 @@
+      uint head = slotblockcache->heads[ txn_bucket ];
+      txnpage->txns[ txn_idx ]->slotblockcache_next = head;
+      FD_COMPILER_MFENCE();
+      if( FD_LIKELY( FD_ATOMIC_CAS( &slotblockcache->heads[ txn_bucket ], head, FD_TXNCACHE_TXNS_PER_PAGE*txnpage_idx+txn_idx )==head ) ) break;
+      FD_SPIN_PAUSE();
+    }
+
+    for(;;) {
+      ulong lowest_slot = blockcache->lowest_slot;
+      if( FD_UNLIKELY( txn->slot>=lowest_slot ) ) break;
+      if( FD_LIKELY( FD_ATOMIC_CAS( &blockcache->lowest_slot, lowest_slot, txn->slot )==lowest_slot ) ) break;
+      FD_SPIN_PAUSE();
+    }
+    return 1;
+  }
+}
+
+int
+fd_txncache_insert_batch( fd_txncache_t *              tc,
+                          fd_txncache_insert_t const * txns,
+                          ulong                        txns_cnt ) {
+  fd_rwlock_read( tc->lock );
+
+  for( ulong i=0UL; i<txns_cnt; i++ ) {
+    fd_txncache_private_blockcache_t * blockcache;
+    if( FD_UNLIKELY( !fd_txncache_ensure_blockcache( tc, txns[ i ].blockhash, &blockcache ) ) ) goto unlock_fail;
+
+    fd_txncache_private_slotcache_t * slotcache;
+    if( FD_UNLIKELY( !fd_txncache_ensure_slotcache( tc, txns[ i ].slot, &slotcache ) ) ) goto unlock_fail;
+
+    fd_txncache_private_slotblockcache_t * slotblockcache;
+    if( FD_UNLIKELY( !fd_txncache_ensure_slotblockcache( slotcache, txns[ i ].blockhash, &slotblockcache ) ) ) goto unlock_fail;
+
+    for(;;) {
+      fd_txncache_private_txnpage_t * txnpage = fd_txncache_ensure_txnpage( tc, blockcache );
+      if( FD_UNLIKELY( !txnpage ) ) goto unlock_fail;
+
+      int success = fd_txncache_insert_txn( tc, blockcache, slotblockcache, txnpage, &txns[ i ] );
+      if( FD_LIKELY( success ) ) break;
+      FD_SPIN_PAUSE();
+    }
+  }
+
+  fd_rwlock_unread( tc->lock );
+  return 1;
+
+unlock_fail:
+  fd_rwlock_unread( tc->lock );
+  return 0;
+}
+
+void
+fd_txncache_query_batch( fd_txncache_t *             tc,
+                         fd_txncache_query_t const * queries,
+                         ulong                       queries_cnt,
+                         void *                      query_func_ctx,
+                         int ( * query_func )( ulong slot, void * ctx ),
+                         int *                       out_results ) {
+  fd_rwlock_read( tc->lock );
+
+  for( ulong i=0UL; i<queries_cnt; i++ ) {
+    out_results[ i ] = 0;
+
+    fd_txncache_query_t const * query = &queries[ i ];
+    fd_txncache_private_blockcache_t * blockcache;
+    int result = fd_txncache_find_blockhash( tc, query->blockhash, &blockcache );
+    if( FD_UNLIKELY( result!=FD_TXNCACHE_FIND_FOUND ) ) {
+      continue;
+    }
+
+    ulong txnhash_offset = blockcache->txnhash_offset;
+    ulong head_hash = FD_LOAD( ulong, query->txnhash+txnhash_offset ) % FD_TXNCACHE_BLOCKCACHE_MAP_CNT;
+    for( uint head=blockcache->heads[ head_hash ]; head!=UINT_MAX; head=tc->txnpages[ head/FD_TXNCACHE_TXNS_PER_PAGE ].txns[ head%FD_TXNCACHE_TXNS_PER_PAGE ]->blockcache_next ) {
```

Comment:
You could consider a macro for `tc->txnpages[ head/FD_TXNCACHE_TXNS_PER_PAGE ].txns[ head%FD_TXNCACHE_TXNS_PER_PAGE ]` since it shows up several times in places it would be nice if it weren't so long